### PR TITLE
RAC-375: dropdown component for DSM

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    {args => {
+      const [isOpen, setOpen] = useState(false);
+      const open = () => setOpen(true);
+      const close = () => setOpen(false);
+      return (
+        <SpaceContainer height={200}>
+          <Dropdown {...args} style={{left: '800px', position: 'absolute'}}>
+            <Dropdown.Action onClick={open}>
+              <Button>yolo</Button>
+            </Dropdown.Action>
```
Comment (tseho) [inline akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:39]:
```suggestion
              <Button onClick={open}>yolo</Button>
```
Do we still need Dropdown.Action ?

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            {isOpen && (
+              <>
+                <Dropdown.Backdrop onClick={close} />
```
Comment (tseho) [inline akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:42]: Do we also need to expose the backdrop ?
Can't we add `onClose` on the root `<Dropdown>` ?

Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:53]: Yes, good idea, I will ask the guys about it :)

File: akeneo-design-system/src/hooks/useShortcut.ts [hunk 3]
```diff
@@ -15,7 +15,8 @@ const useShortcut = <NodeType extends HTMLElement>(
 ): Ref<NodeType> => {
   const memoizedCallback = useCallback(
     (event: KeyboardEvent) => {
-      if (key === event.code) {
+      if (key === event.key) {
+        event.stopImmediatePropagation();
```
Comment (ValentinMumble) [inline akeneo-design-system/src/hooks/useShortcut.ts:19]: I think @tseho made a specific PR about this, should we or should we not stop propagation? 🤔

File: akeneo-design-system/src/storybook/test-util.tsx [hunk 4]
```diff
@@ -7,4 +7,4 @@ const wrapper: ComponentType = ({children}) => <ThemeProvider theme={pimTheme}>{
 const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) => render(ui, {wrapper, ...options});
 
 export * from '@testing-library/react';
-export {customRender as render};
+export {customRender as render, wrapper};
```
Comment (ValentinMumble) [inline akeneo-design-system/src/storybook/test-util.tsx:10]: Why?

File: akeneo-design-system/src/hooks/useShortcut.ts (hunk 3 above)
Comment (juliensnz) [reply akeneo-design-system/src/hooks/useShortcut.ts:19]: Yes, it's a merge fail I think :)

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/hooks/useShortcut.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import {useCallback, useEffect} from 'react';
 import {Key} from '../tools';
 
 const useShortcut = (key: Key, callback: Function) => {
-  const memoizedCallback = useCallback((event: KeyboardEvent) => (key === event.code ? callback(event) : null), [
+  const memoizedCallback = useCallback((event: KeyboardEvent) => (key === event.key ? callback(event) : null), [
```
Comment (ValentinMumble) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/hooks/useShortcut.ts:5]: I trashed this file in another PR, and replaced usages with the DSM one

File: akeneo-design-system/src/storybook/test-util.tsx (hunk 4 above)
Comment (juliensnz) [reply akeneo-design-system/src/storybook/test-util.tsx:10]: Yes, I think I tried a lot of things during testing and forgot to remove it.

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/hooks/useShortcut.ts (hunk 5 above)
Comment (juliensnz) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/hooks/useShortcut.ts:5]: ok, perfect, I reverse the change :)

File: akeneo-design-system/src/components/index.ts [hunk 6]
```diff
@@ -15,3 +15,5 @@ export * from './MessageBar/MessageBar';
 export * from './Modal/Modal';
 export * from './ProgressBar/ProgressBar';
 export * from './Table/Table';
+export * from './Input/BooleanInput/BooleanInput';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/index.ts:18]:
```suggestion
```
BooleanInput is already exported from ./Input line 10

File: akeneo-design-system/src/components/Dropdown/ItemCollection/ItemCollection.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import {Key, Override} from '../../../shared';
+import React, {ReactNode, useRef, Ref, useCallback, KeyboardEvent} from 'react';
+import styled from 'styled-components';
+import {useAutoFocus} from 'hooks';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/ItemCollection/ItemCollection.tsx:4]:
```suggestion
import {useAutoFocus} from '../../../hooks';
```

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
+import {useShortcut} from 'hooks';
+import React, {ReactNode, useRef, useState, useEffect} from 'react';
+import {Key} from '../../../shared/key';
+import styled, {css} from 'styled-components';
+import {AkeneoThemedProps, getColor} from 'theme';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:5]:
```suggestion
import {AkeneoThemedProps, getColor} from '../../../theme';
```

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import {useShortcut} from 'hooks';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:1]:
```suggestion
import {useShortcut} from '../../../hooks';
```

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import React, {ReactNode, useRef, useState, useEffect} from 'react';
+import {Key} from '../../../shared/key';
+import styled, {css} from 'styled-components';
+import {AkeneoThemedProps, getColor} from 'theme';
+
+const Container = styled.div<
+  {
+    visible: boolean;
+    verticalPosition: 'up' | 'down';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:10]: Maybe extract to type as it is used several times? (line 42, 49)

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+      ? css`
+          right: -1px;
+        `
+      : css`
+          left: -1px;
+        `};
+`;
+
+type OverlayProps = {
+  position?: 'up' | 'down';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:42]: No JSDoc?

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  onClose: () => void;
+  children: ReactNode;
+};
+
+const Overlay = ({position, onClose, children}: OverlayProps) => {
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:47]:
```suggestion
const Overlay = ({position = 'down', onClose, children}: OverlayProps) => {
```

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.unit.tsx [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+  fireEvent.keyDown(window, {key: 'Escape', code: 'Escape'});
+
+  expect(onClose).toBeCalledTimes(1);
+});
+
+test('it guesses the position best suited to display the Overlay', () => {
+  const onClose = jest.fn();
+  render(
+    <ThemeProvider theme={pimTheme}>
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.unit.tsx:21]: Do you really need the ThemeProvider ? Here it's should be provided by the render method

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+      : css`
+          left: -1px;
+        `};
+`;
+
+type OverlayProps = {
+  /**
+   * Vertical position of the overlay (forced)
+   */
+  position?: VerticalPosition;
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:48]:
```suggestion
  verticalPosition?: VerticalPosition;
```

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+      if (undefined === position) {
+        const elementHeight = overlayRef.current.getBoundingClientRect().height;
+        const windowHeight = window.innerHeight;
+        const distanceToTop = overlayRef.current.getBoundingClientRect().top;
+        const distanceToBottom = windowHeight - (elementHeight + distanceToTop);
+
+        setVerticalPosition(distanceToTop > distanceToBottom ? 'up' : 'down');
+      }
+
+      if (null !== overlayRef.current) {
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:76]: Already checked on line 66 should we need to re-check ?

File: akeneo-design-system/src/components/Dropdown/Item/Item.unit.tsx [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+      <Item>
+        <Image src="" alt="A nice Image" />
+      </Item>
+    </>
+  );
+
+  expect(screen.getByAltText('A nice Image')).toBeInTheDocument();
+  expect(screen.getByAltText('A nice Image')).toHaveProperty('width', 34);
+  expect(screen.getByAltText('A nice Image')).toHaveProperty('height', 34);
+  expect(screen.getByAltText('A nice Image').parentNode).toHaveStyleRule('height', '44px');
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Item/Item.unit.tsx:33]: In my opinion this should be visually tested not unitary

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+  {
+    visible: boolean;
+    verticalPosition: VerticalPosition;
+    horizontalPosition: HorizontalPosition;
+  } & AkeneoThemedProps
+>`
+  background: ${getColor('white')};
+  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.3);
+  padding: 0 0 10px 0;
+  max-width: 320px;
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:20]: Maybe a max height for long list ?

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+import {Button, Image, Badge, Link, Checkbox} from '../../components';
+import {useBooleanState} from '../../hooks';
+import {useState} from 'react';
+import {SpaceContainer, useSelection} from '../../storybook/PreviewGallery.tsx';
+
+<Meta
+  title="Components/Dropdown"
+  component={Dropdown}
+  argTypes={{
+    level: {control: {type: 'select', options: ['primary', 'warning', 'danger']}},
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:12]: Level is not used i think you can remove it

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+        const elementWidth = overlayRef.current.getBoundingClientRect().width;
+        const windowWidth = window.innerWidth;
+        const distanceToLeft = overlayRef.current.getBoundingClientRect().left;
+        const distanceToRight = windowWidth - (elementWidth + distanceToLeft);
+
+        setHorizontalPosition(distanceToLeft > distanceToRight ? 'left' : 'right');
+      }
+      setVisible(true);
+    }
+  }, []);
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:86]: It depend on position ?
```suggestion
  }, [position]);
```

File: akeneo-design-system/src/components/Dropdown/Item/Item.tsx [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    }, []);
+    const handleKeyDown = useCallback((event: KeyboardEvent<HTMLDivElement>) => {
+      if (Key.Enter === event.key || Key.Space === event.key) {
+        event.preventDefault();
+        handleClick();
+        return;
+      }
+
+      onKeyDown && onKeyDown(event);
+    }, []);
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Dropdown/Item/Item.tsx:71]: ?

```suggestion
    }, [handleClick, onKeyDown]);
```

**Review by StevenVAIDIE — APPROVED** [review]

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx (hunk 14 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:48]: Yes, that is what I was thinking about more and more. I will change that :)

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx (hunk 15 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:76]: Typescript is complaining :/

File: akeneo-design-system/src/components/Dropdown/Item/Item.unit.tsx (hunk 16 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Item/Item.unit.tsx:33]: Yes, I can remove them :)

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx (hunk 18 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:12]: oups!

File: akeneo-design-system/src/components/Dropdown/Item/Item.tsx (hunk 20 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Item/Item.tsx:71]: yes, you are right

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx (hunk 17 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Overlay/Overlay.tsx:20]: The max height is on the Item collection :)

File: akeneo-design-system/src/components/Dropdown/Overlay/Overlay.unit.tsx (hunk 13 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Overlay/Overlay.unit.tsx:21]: removed!

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx [hunk 21]
```diff
@@ -0,0 +1,181 @@
+import {Meta, Story, ArgsTable, Canvas} from '@storybook/addon-docs/blocks';
+import {Dropdown} from './Dropdown.tsx';
+import {Button, Image, Badge, Link, Checkbox} from '../../components';
+import {useBooleanState} from '../../hooks';
+import {useState} from 'react';
+import {SpaceContainer, useSelection} from '../../storybook/PreviewGallery.tsx';
+
+<Meta title="Components/Dropdown" component={Dropdown} />
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:8]: Maybe we can the subcomponents here as in the Table, to have their props documented as well, what do you think?
image

File: akeneo-design-system/src/components/Dropdown/Dropdown.visual.tsx [hunk 22]
```diff
@@ -0,0 +1,23 @@
+import 'expect-puppeteer';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Dropdown.visual.tsx:1]: Should we add the Dropdown to the ignored ones in `all.visual.tsx` as it is pretty useless as is?

Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Dropdown.visual.tsx:1]: Yes, you are right

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx (hunk 21 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:8]: Yes, could be a good idea. I will try to do it today!

File: akeneo-design-system/src/components/Dropdown/Item/Item.unit.tsx [hunk 23]
```diff
@@ -0,0 +1,71 @@
+import React from 'react';
+import {Item} from './Item';
+import {Image, Link, Checkbox} from '../../../components';
+import {render, screen, fireEvent} from '../../../storybook/test-util';
+import 'jest-styled-components';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/Item/Item.unit.tsx:5]:
```suggestion
```
This is unneeded now that they are tested visually no?

File: akeneo-design-system/src/components/Dropdown/ItemCollection/ItemCollection.unit.tsx [hunk 24]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import {ItemCollection} from './ItemCollection';
+import {render, screen, fireEvent} from '../../../storybook/test-util';
+import 'jest-styled-components';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Dropdown/ItemCollection/ItemCollection.unit.tsx:4]:
```suggestion
```

**Review by ValentinMumble — APPROVED** [review]
Missing disabled item state but GTMMMMMMMM for now

File: akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx (hunk 21 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Dropdown/Dropdown.stories.mdx:8]: This is currently failing. So we postone it to the next PR

**Pull request merged**